In [33]:
# standard library
import os
import pandas as pd
import pandas_gbq as gbq
import numpy as np
import requests
from dotenv import load_dotenv
load_dotenv(dotenv_path="../config/.env")


# datetime and timezone library
from datetime import datetime, timezone
from pytz import timezone as tz

# geocoding library
from geopy.geocoders import Nominatim
from geopy.extra.rate_limiter import RateLimiter

# Google API libraries
from google.oauth2 import service_account

# Logging
import logging
# setup logging
logging.basicConfig(level=logging.INFO)


## Extracting Data from Open Weather Map


In [9]:
# 1. Load .env dan ambil API Key
from dotenv import load_dotenv
import os
load_dotenv(dotenv_path="../config/.env")

API_KEY = os.getenv("OWM_API_KEY")
print("API KEY FOUND:", bool(API_KEY))

API KEY FOUND: True


In [10]:
# Define Base URL
API_KEY = os.getenv("OWM_API_KEY")
BASE_URL = "https://api.openweathermap.org/data/2.5/weather"
CITY_LIST_PATH = "../config/city_list.csv"

In [11]:
def fetch_weather_by_id(city_id: int) -> dict:
    """Fetch weather data by OpenWeatherMap city ID"""
    params = {
        "id": city_id,
        "appid": API_KEY,
        "units": "metric"
    }
    try:
        response = requests.get(BASE_URL, params=params)
        response.raise_for_status()
        data = response.json()
        return {
            "city_id": data["id"],
            "city": data["name"],
            "country": data["sys"]["country"],
            "lat": data["coord"]["lat"],
            "lon": data["coord"]["lon"],
            "temperature": data["main"]["temp"],
            "humidity": data["main"]["humidity"],
            "weather": data["weather"][0]["main"],
            "description": data["weather"][0]["description"],
            "wind_speed": data["wind"]["speed"],
            "timestamp": datetime.utcfromtimestamp(data["dt"]),
            "fetched_at": datetime.utcnow()
        }
    except Exception as e:
        logging.error(f"Error fetching city ID {city_id}: {e}")
        return None

def fetch_weather_all_cities(city_file: str = CITY_LIST_PATH) -> pd.DataFrame:
    cities_df = pd.read_csv(city_file)
    records = []
    for _, row in cities_df.iterrows():
        result = fetch_weather_by_id(row["id"])
        if result:
            records.append(result)
        else:
            logging.warning(f"Skipping city ID {row['id']} ({row['name']})")
    return pd.DataFrame(records)


In [ ]:
data = fetch_weather_all_cities()

In [14]:
df = data
df

city_id          city country      lat       lon  temperature  humidity  \
0   1650077    Banyuwangi      ID  -8.2325  114.3576        24.24        95   
1   1625822      Surabaya      ID  -7.2492  112.7508        30.00        78   
2   1835848         Seoul      KR  37.5683  126.9778        30.22        71   
3   5128581      New York      US  40.7143  -74.0060        21.60        84   
4    292223         Dubai      AE  25.2582   55.3047        41.97        35   
5   1850147         Tokyo      JP  35.6895  139.6917        29.03        70   
6    360630         Cairo      EG  30.0626   31.2497        40.75        10   
7   1273294         Delhi      IN  28.6667   77.2167        35.22        36   
8   3435910  Buenos Aires      AR -34.6132  -58.3772         9.39        81   
9   1816670       Beijing      CN  39.9075  116.3972        28.77        90   
10  3448439     São Paulo      BR -23.5475  -46.6361        15.99        85   
11  1796236      Shanghai      CN  31.2222  121.4581        28.68        86   
12   745044      Istanbul      TR  41.0138   28.9497        34.43        35   
13  2968815         Paris      FR  48.8534    2.3486        21.36        70   
14  3369157     Cape Town      ZA -33.9258   18.4232        16.23        52   
15  2147714        Sydney      AU -33.8679  151.2073        12.66        68   
16   993800  Johannesburg      ZA -26.2023   28.0436        15.21        13   
17  1853909         Osaka      JP  34.6937  135.5022        30.34        71   
18  1264527       Chennai      IN  13.0878   80.2785        33.77        60   
19   344979   Addis Ababa      ET   9.0250   38.7469        20.58        62   
20   745042      Istanbul      TR  41.0351   28.9833        34.43        35   
21  6167865       Toronto      CA  43.7001  -79.4163        22.75        92   
22  2643743        London      GB  51.5085   -0.1257        20.10        58   
23  1701668        Manila      PH  14.6042  120.9822        27.20        94   
24  3871336      Santiago      CL -33.4569  -70.6483         9.79        85   
25   524901        Moscow      RU  55.7522   37.6156        27.72        39   
26  1277333     Bengaluru      IN  12.9762   77.6033        26.70        71   
27  3067696        Prague      CZ  50.0880   14.4208        21.74        57   
28  1856057        Nagoya      JP  35.1815  136.9064        29.80        67   
29  1261481     New Delhi      IN  28.6128   77.2311        35.09        36   
30  1733046  Kuala Lumpur      MY   3.1431  101.6865        30.79        70   
31  1880252     Singapore      SG   1.2897  103.8501        30.78        65   
32  6160752        Sydney      CA  46.1501  -60.1818        17.21        61   
33  2174003      Brisbane      AU -27.4679  153.0281        13.23        65   

   weather       description  wind_speed           timestamp  \
0   Clouds  scattered clouds        1.14 2025-07-27 11:25:57   
1   Clouds        few clouds        3.96 2025-07-27 11:22:32   
2    Clear         clear sky        1.96 2025-07-27 11:20:57   
3   Clouds   overcast clouds        3.08 2025-07-27 11:24:47   
4   Clouds  scattered clouds        7.12 2025-07-27 11:16:34   
5   Clouds   overcast clouds        8.62 2025-07-27 11:17:55   
6    Clear         clear sky        5.30 2025-07-27 11:23:00   
7   Clouds   overcast clouds        2.10 2025-07-27 11:26:07   
8   Clouds   overcast clouds        7.12 2025-07-27 11:22:35   
9   Clouds   overcast clouds        2.39 2025-07-27 11:20:40   
10  Clouds        few clouds        1.75 2025-07-27 11:24:24   
11  Clouds  scattered clouds        4.08 2025-07-27 11:21:29   
12   Clear         clear sky        4.94 2025-07-27 11:21:37   
13  Clouds   overcast clouds        3.73 2025-07-27 11:26:12   
14   Clear         clear sky        1.95 2025-07-27 11:22:26   
15  Clouds     broken clouds        7.41 2025-07-27 11:24:21   
16   Clear         clear sky        0.96 2025-07-27 11:21:16   
17  Clouds     broken clouds        5.46 2025-07-27 11:24:31   
18  Clouds   overcast 

In [17]:
df.describe()

city_id        lat         lon  temperature   humidity  \
count  3.400000e+01  34.000000   34.000000    34.000000  34.000000   
mean   2.154912e+06  16.801215   54.171185    25.522647  63.147059   
min    2.922230e+05 -34.613200  -79.416300     9.390000  10.000000   
25%    1.266719e+06  -5.114475   15.421400    20.220000  53.250000   
50%    1.826259e+06  28.639750   66.260700    27.460000  67.500000   
75%    2.887547e+06  40.512600  115.887300    30.670000  80.250000   
max    6.167865e+06  55.752200  153.028100    41.970000  95.000000   
std    1.488930e+06  29.323654   71.656588     8.497247  22.238558   

       wind_speed                      timestamp  \
count   34.000000                             34   
mean     3.870588  2025-07-27 11:22:30.647058944   
min      0.750000            2025-07-27 11:13:14   
25%      1.952500  2025-07-27 11:21:01.750000128   
50%      3.265000     2025-07-27 11:22:47.500000   
75%      5.337500  2025-07-27 11:25:48.750000128   
max      8.950000            2025-07-27 11:26:28   
std      2.439558                            NaN   

                          fetched_at  
count                             34  
mean   2025-07-27 11:26:16.124194304  
min       2025-07-27 11:25:58.255875  
25%    2025-07-27 11:26:09.479122688  
50%    2025-07-27 11:26:16.306757376  
75%    2025-07-27 11:26:24.080024064  
max       2025-07-27 11:26:30.577454  
std                              NaN

# Descriptive Analysis - OpenWeatherMap ETL Pipeline

This document summarizes descriptive insights gathered from a single ETL pipeline execution fetching real-time weather data from 50 global cities using the OpenWeatherMap API.

## Purpose of ETL Pipeline

The pipeline is designed to automate the collection of real-time weather data from multiple global cities, clean and enrich it, store it in Google BigQuery, and present it through an interactive dashboard. This supports analysis, monitoring, and visualization of global weather conditions on a recurring basis.

---

## Descriptive Insights

### 1. Distribution of Weather Conditions

- Most cities during the snapshot were experiencing **Clouds** or **Clear** skies.
- A few cities were reporting **Rain** or **Moderate Rain** (e.g., Addis Ababa, Manila).

**Interpretation:** Globally, the dominant real-time weather pattern was overcast or clear, with localized rainfall events.

---

### 2. Temperature Extremes

- **Highest Temperatures:**
  - Cairo (40.75°C)
  - Dubai (41.97°C)
  - Delhi (37.6°C)

- **Lowest Temperatures:**
  - Cape Town (16.23°C)
  - Sydney (16.03°C)
  - São Paulo (15.99°C)

**Interpretation:** Cities in arid or subtropical zones recorded significantly higher temperatures, supporting potential alert classification.

---

### 3. Temperature Categories

(Assuming enrichment step added a `temp_category` column.)

- **Hot:** Cairo, Dubai, Delhi
- **Warm:** Jakarta, Tokyo, New York
- **Cold:** Cape Town, São Paulo, Sydney

**Interpretation:** This classification allows for geographical grouping and temporal comparison across cities.

---

### 4. Notable Wind Conditions

- Cities such as **Osaka**, **Bengaluru**, and **Istanbul** recorded wind speeds above 8 m/s.

**Interpretation:** These cities may be experiencing wind-driven weather activity, which is relevant for alerts and infrastructure planning.

---

### 5. Combined Conditions (Risk Zones)

- **Cairo** and **Dubai**: High temperatures with clear skies (potential heatwave zones)
- **Manila**: High temperature + rain (high humidity zone)

---

## Why These Insights Matter

These insights justify the creation of the ETL pipeline:

- **Without automation**, gathering this scale of data daily would be unfeasible.
- **Enables trend detection**, anomaly identification, and support for weather-sensitive decisions (e.g., agriculture, disaster readiness, transportation).
- **Scalable foundation** for adding alerting, forecasting, or geospatial mapping in the future.

---

## Conclusion

The descriptive analysis demonstrates that even a single ETL run can offer a valuable snapshot of global weather variation. With continuous automation, the data becomes increasingly powerful for monitoring trends and supporting decision-making.

## Transforming and Enriching the Data

In [ ]:
# Default: Waktu Jakarta (bisa disesuaikan)
WIB = timezone("Asia/Jakarta")

def clean_weather_data(df: pd.DataFrame) -> pd.DataFrame:
    """
    Bersihkan dan transformasi data cuaca:
    - Pastikan tipe data konsisten
    - Drop null penting
    - Konversi timestamp ke timezone lokal
    """
    if df.empty:
        return df

    df = df.dropna(subset=["temperature", "humidity", "weather", "timestamp"])

    df["timestamp"] = pd.to_datetime(df["timestamp"]).dt.tz_localize("UTC").dt.tz_convert(WIB)
    df["fetched_at"] = pd.to_datetime(df["fetched_at"]).dt.tz_localize("UTC").dt.tz_convert(WIB)

    return df


def enrich_weather_data(df: pd.DataFrame) -> pd.DataFrame:
    """
    Tambahkan kolom turunan:
    - temp_category (dingin, hangat, panas)
    - local_time, day_of_week, hour_of_day
    """
    if df.empty:
        return df

    def classify_temp(temp):
        if temp < 20:
            return "Cold"
        elif temp < 30:
            return "Warm"
        else:
            return "Hot"

    df["temp_category"] = df["temperature"].apply(classify_temp)
    df["day_of_week"] = df["timestamp"].dt.day_name()
    df["hour_of_day"] = df["timestamp"].dt.hour

    return df

clean = clean_weather_data(df)
enriched = enrich_weather_data(clean)

In [16]:
enriched.head()

city_id        city country      lat       lon  temperature  humidity  \
0  1650077  Banyuwangi      ID  -8.2325  114.3576        24.24        95   
1  1625822    Surabaya      ID  -7.2492  112.7508        30.00        78   
2  1835848       Seoul      KR  37.5683  126.9778        30.22        71   
3  5128581    New York      US  40.7143  -74.0060        21.60        84   
4   292223       Dubai      AE  25.2582   55.3047        41.97        35   

  weather       description  wind_speed                 timestamp  \
0  Clouds  scattered clouds        1.14 2025-07-27 18:25:57+07:00   
1  Clouds        few clouds        3.96 2025-07-27 18:22:32+07:00   
2   Clear         clear sky        1.96 2025-07-27 18:20:57+07:00   
3  Clouds   overcast clouds        3.08 2025-07-27 18:24:47+07:00   
4  Clouds  scattered clouds        7.12 2025-07-27 18:16:34+07:00   

                        fetched_at temp_category day_of_week  hour_of_day  
0 2025-07-27 18:25:58.255875+07:00          Warm      Sunday           18  
1 2025-07-27 18:25:59.267957+07:00           Hot      Sunday           18  
2 2025-07-27 18:26:00.351718+07:00           Hot      Sunday           18  
3 2025-07-27 18:26:03.542518+07:00          Warm      Sunday           18  
4 2025-07-27 18:26:04.268195+07:00           Hot      Sunday           18

In [19]:
enriched.describe()  

city_id        lat         lon  temperature   humidity  \
count  3.400000e+01  34.000000   34.000000    34.000000  34.000000   
mean   2.154912e+06  16.801215   54.171185    25.522647  63.147059   
std    1.488930e+06  29.323654   71.656588     8.497247  22.238558   
min    2.922230e+05 -34.613200  -79.416300     9.390000  10.000000   
25%    1.266719e+06  -5.114475   15.421400    20.220000  53.250000   
50%    1.826259e+06  28.639750   66.260700    27.460000  67.500000   
75%    2.887547e+06  40.512600  115.887300    30.670000  80.250000   
max    6.167865e+06  55.752200  153.028100    41.970000  95.000000   

       wind_speed  hour_of_day  
count   34.000000         34.0  
mean     3.870588         18.0  
std      2.439558          0.0  
min      0.750000         18.0  
25%      1.952500         18.0  
50%      3.265000         18.0  
75%      5.337500         18.0  
max      8.950000         18.0

# Additional Descriptive Insights — Enriched Weather Data (Sample of 5 Cities)

## 1. Distribution of Temperature Categories

- **Hot:** Surabaya, Seoul, Dubai  
- **Warm:** Banyuwangi, New York

**Interpretation:**  
Despite all cities being observed at the same hour (`hour_of_day = 18`), thermal conditions vary significantly. This highlights the importance of location-based temperature classification for global comparative analysis.

---

## 2. Humidity Extremes

- **Banyuwangi** recorded the highest humidity at **95%**
- **Dubai**, despite having the highest temperature (41.97°C), had the lowest humidity at **35%**

**Interpretation:**  
Such variations are relevant for calculating heat index or thermal discomfort zones. This can be a valuable enrichment for future transformations.

---

## 3. Weather Conditions at the Same Hour

All records are from **18:00 local time**, yet:
- Most cities had **cloud-related conditions** (`Clouds`, `scattered clouds`)
- **Seoul** had a clear sky

**Interpretation:**  
This supports time-based weather pattern analysis and can be used for simple hourly trend prediction models.

---

## 4. Wind Speed Variation

- **Dubai** had the highest wind speed at **7.12 m/s**
- **Banyuwangi** and **Seoul** had lower wind speeds below 2 m/s

**Interpretation:**  
Wind speed variation contributes to weather severity classification or comfort index analysis.

---

## 5. Temporal Consistency

- All data points were fetched on **Sunday** at **hour 18**
- This confirms temporal alignment of the dataset, which is crucial for time-series analysis and consistent dashboard visuals

---

## Recommendations

- Derive a `heat_index` field from temperature and humidity
- Analyze `temp_category` trends across hours to identify daily thermal cycles
- Create scatter plots of `humidity vs temperature` to detect thermal outliers


In [21]:
df.to_csv(f'../data/weather-daily-{datetime.now().strftime("%Y-%m-%d")}.csv', index=False)

In [36]:
# BigQuery Configuration
PROJECT_ID = os.getenv("PROJECT_ID")
TABLE_ID = os.getenv("TABLE_ID")
CREDENTIALS_PATH = "../config/owm-etl-project-986252b4609c.json"

In [37]:
from pandas_gbq import to_gbq
def upload_to_bigquery(
    df: pd.DataFrame,
    project_id: str = PROJECT_ID,
    table_id: str = TABLE_ID,
    credentials_path: str = CREDENTIALS_PATH,
    if_exists: str = "replace"
):
    """
    Unggah DataFrame ke Google BigQuery.

    Parameters:
        df (pd.DataFrame): Data yang akan diunggah
        project_id (str): ID project GCP
        table_id (str): Format dataset.table
        credentials_path (str): Path ke JSON file service account
        if_exists (str): Mode upload: 'replace', 'append', 'fail'
    """
    try:
        credentials = service_account.Credentials.from_service_account_file(credentials_path)
        logging.info(f"Uploading to BigQuery: {table_id} (mode: {if_exists})")
        to_gbq(
            dataframe=df,
            destination_table=table_id,
            project_id=project_id,
            credentials=credentials,
            if_exists=if_exists
        )
        logging.info("✅ Upload success.")
    except Exception as e:
        logging.error(f"❌ Upload failed: {e}")

In [38]:
upload_to_bigquery(enriched, if_exists="append")

INFO:root:Uploading to BigQuery: weather_data.daily (mode: append)
34 out of 34 rows loaded.ctor:
INFO:root:✅ Upload success.
